src/backend/storage/lmgr/lock.c

```diff
@@ -4264,3 +4270,25 @@ VirtualXactLock(VirtualTransactionId vxid, bool wait)
 	LockRelease(&tag, ShareLock, false);
 	return true;
 }
+
+static void
+setFPHoldTillEndXact(Oid relid)
+{
+	uint32  f;
+	PGPROC *proc = MyProc;
+
+	LWLockAcquire(proc->backendLock, LW_EXCLUSIVE);
+
+	for (f = 0; f < FP_LOCK_SLOTS_PER_BACKEND; ++f)
+	{
+		uint32 lockbits = FAST_PATH_GET_BITS(proc, f);
+
+		if (!lockbits ||
+			proc->fpRelId[f] != relid)
+			continue;
+
+		proc->fpHoldTillEndXact[f] = true;
```

Comment:
If we have found the lock in fastpath could we break now?

Reply:
I have considered this. But I am not sure whether the remaining array contains a relation lock for the same relid. @guofengrichard What is your idea?

Reply:
A `break` here is good since one relid only occupies one slot.
In addition, suggest to use the following pattern to look for the matched slot, as most time there is no necessary to get the lock bits.
```
 if (relid != proc->fpRelId[f] || FAST_PATH_GET_BITS(proc, f) == 0)
                continue;
```

Reply:
@guofengrichard Thanks. I will improve the code.